linux/keyman-config/keyman_config/convertico.py

```diff
@@ -4,42 +4,102 @@
-
+Image.LOAD_TRUNCATED_IMAGES = True
 
 def changeblacktowhite(im):
-    data = np.array(im)   # "data" is a height x width x 4 numpy array
-    red, green, blue, alpha = data.T # Temporarily unpack the bands for readability
+	data = np.array(im)   # "data" is a height x width x 4 numpy array
+	red, green, blue, alpha = data.T # Temporarily unpack the bands for readability
+
+	# Replace black with white... (leaves alpha values alone...)
+	white_areas = (red == 0) & (blue == 0) & (green == 0)
+	data[..., :-1][white_areas.T] = (255, 255, 255)  # Transpose back needed
 
-    # Replace black with white... (leaves alpha values alone...)
-    white_areas = (red == 0) & (blue == 0) & (green == 0)
-    data[..., :-1][white_areas.T] = (255, 255, 255) # Transpose back needed
+	im2 = Image.fromarray(data)
+	return im2
 
-    im2 = Image.fromarray(data)
-    return im2
 
 def checkandsaveico(icofile):
-    im = Image.open(icofile)
-    im = im.convert('RGBA')
-    im2 = im
-    num, colour = max(im.getcolors(im.size[0]*im.size[1]))
-    logging.debug("checkandsaveico maxcolour: num {0}: colour {1}".format(num, colour))
-    if num > 160 and colour == (0, 0, 0, 0):
-        logging.info("checkandsaveico:" + icofile + " mostly black so changing black to white")
-        im2 = changeblacktowhite(im)
-    im2.save(icofile + ".bmp")
-    im3 = Image.open(icofile + ".bmp")
-    im4 = im3.resize((64, 64), Image.ANTIALIAS)
-    im4.save(icofile + ".png")
-    os.remove(icofile + ".bmp")
+	"""
+	Convert keyman icofile to 64x64 png to work in IBus
+	The starting icofile may be ico or bmp format
+
+	Args:
+		icofile (str): path to ico file
+	"""
+	name, ext = os.path.splitext(icofile)
+	bmpfile = name + ".bmp"
+	if ext == '.ico':
+		im = Image.open(icofile)
+		im = im.convert('RGBA')
+		im2 = im
+		num, colour = max(im.getcolors(im.size[0]*im.size[1]))
+		logging.debug("checkandsaveico maxcolour: num {0}: colour {1}".format(num, colour))
+		if num > 160 and colour == (0, 0, 0, 0):
+			logging.info("checkandsaveico:" + icofile + " mostly black so changing black to white")
+			im2 = changeblacktowhite(im)
+		im2.save(bmpfile)
+
+	try:
+		im3 = Image.open(bmpfile)
+		im4 = im3.resize([64, 64], Image.ANTIALIAS)
+		im4.save(name + '.png', 'png')
+	except (IOError, OSError) as e:
+		logging.error("Cannot convert %s to png", icofile)
+		pass
+
+
+def extractico(kmxfile):
+	"""
+	Extract icon file from compiled kmx keyboard
+
+	Args:
+		kmxfile (str): path to kmx file
+	"""
+	name, ext = os.path.splitext(kmxfile)
+	imagefilename = name
+	with open(kmxfile, mode='rb') as file: # b is important -> binary
+		fileContent = file.read()
+
+		kmxstart = struct.unpack_from("<16I", fileContent, 0)
+		if kmxstart[0] != 0x5354584B:
+			logging.debug("bad kmx identifier")
+			return False
+		bitmapOffset = kmxstart[14]
+		bitmapSize = kmxstart[15]
+		logging.debug("bitmap offset is %d", bitmapOffset)
+		logging.debug("bitmap size is %d", bitmapSize)
+		file.seek(bitmapOffset, 0)
```

Comment:
Can we test for error here, in case the file is invalid?

Reply:
I can add a sanity check after `file.read()`. afaict, `file.seek` doesn't return a status...